tonic/functional/event_downsampling.py

```diff
@@ -0,0 +1,170 @@
+        dt (float): temporal resolution of events in milliseconds.
+        noise_threshold (int): number of events before a spike representing a new event is emitted.
+        differentiator_call (bool): Preserve frame spikes for differentiator method in order to optimise 
+                                    differentiator method.
+        
+    Returns:
+        the downsampled input events using the integrator method.
+    """
+    
+    assert "x" and "y" and "t" in events.dtype.names
+    assert isinstance(noise_threshold, int)
+    assert dt is not None
+    
+    events = events.copy()
+    
+    if np.issubdtype(events["t"].dtype, np.integer):
+        dt *= 1000
+        dt_scaling = True
+    
+    if differentiator_call:
+        assert dt // events["t"][-1] == 0
+    
+    # Downsample
+    spatial_factor = np.asarray(target_size) / sensor_size[:-1]
+
+    events["x"] = events["x"] * spatial_factor[0]
+    events["y"] = events["y"] * spatial_factor[1]
+    
+    # Re-format event times to new temporal resolution
+    events_sliced = slice_events_by_time(events, time_window=dt)
+    
+    # Running buffer of events in each pixel
+    frame_spike = np.zeros(np.flip(target_size))
+    event_histogram = []
+    
+    events_new = []
+    
+    for time, event in enumerate(events_sliced):
+        # Separate by polarity
+        xy_pos = event[event["p"] == 1]
+        xy_neg = event[event["p"] == 0]
+        
+        # Sum in 2D space using histogram
+        frame_histogram = np.subtract(np.histogram2d(xy_pos["y"], xy_pos["x"], [range(target_size[1] + 1), range(target_size[0] + 1)])[0],
+                                      np.histogram2d(xy_neg["y"], xy_neg["x"], [range(target_size[1] + 1), range(target_size[0] + 1)])[0])
+        
+        frame_spike += frame_histogram
+            
+        coordinates_pos = np.stack(np.nonzero(np.maximum(frame_spike >= noise_threshold, 0))).T
+        coordinates_neg = np.stack(np.nonzero(np.maximum(-frame_spike >= noise_threshold, 0))).T
+        
+        if np.logical_or(coordinates_pos.size, coordinates_neg.size).sum():
+        
+            # For optimising differentiator
+            event_histogram.append((time*dt, frame_spike.copy()))
+            
+            # Reset spiking coordinates to zero
+            frame_spike[coordinates_pos[:,0], coordinates_pos[:,1]] = 0
+            frame_spike[coordinates_neg[:,0], coordinates_neg[:,1]] = 0
+            
+            # Restructure events
+            events_new.append(np.column_stack((np.flip(coordinates_pos, axis=1), np.ones((coordinates_pos.shape[0],1)).astype(dtype=bool), 
+                                                (time*dt)*np.ones((coordinates_pos.shape[0],1)))))
+            
+            events_new.append(np.column_stack((np.flip(coordinates_neg, axis=1), np.zeros((coordinates_neg.shape[0],1)).astype(dtype=bool), 
+                                                (time*dt)*np.ones((coordinates_neg.shape[0],1)))))
```

Comment:
can you please explain to me what's happening here? it seems like you're creating frames from events

Reply:
The events are first split into ON and OFF events. The events then form 2D histograms to difference the ON from OFF events. This, I believe is what you're referring to as 'frames'. These histograms are added to a running buffer of 'frame' events that carry the x/y positions of pixel-wise events. These could be ravelled to form an `(1, target_size[0] * target_size[1])` array of events.

Reply:
so why not use the functional call of to_frame here? https://github.com/neuromorphs/tonic/blob/develop/tonic/functional/to_frame.py

I benchmarked np.histogram2d in the past and it's surprisingly slow. Using the existing to_frame you reduce duplicate code and you make sure you get the fastest implementation that I've found. It will give you two channels (one for each polarity). Then you can simply subtract them, no?

Reply:
I too have found np.histogram2d to be slow. However, using ToFrame to me to replicate histogram2d's behaviour did not occur to me. I have now fixed that.